lib/rules/no-restricted-imports.js

```diff
@@ -121,31 +139,128 @@ module.exports = {
         }
 
         /**
+         * Report a restricted path specifically for patterns.
+         * @param {node} node - representing the restricted path reference
+         * @returns {void}
+         * @private
+         */
+        function reportPathForPatterns(node) {
+            const importName = node.source.value.trim();
+
+            context.report({
+                node,
+                message: "'{{importName}}' import is restricted from being used by a pattern.",
+                data: {
+                    importName
+                }
+            });
+        }
+
+        /**
+         * Report a restricted path specifically for patterns.
+         * @param {string} name - path of the import
+         * @param {node} node - representing the restricted path reference
+         * @returns {void}
+         * @private
+         */
+        function reportPathForEverythingImported(name, node) {
+            const importName = node.source.value.trim();
+            const importObjects = restrictedPathMessages[name].objectNames;
+
+            context.report({
+                node,
+                message: "* import is invalid because '{{importObjects}}' from '{{importName}}' is restricted.",
+                data: {
+                    importName,
+                    importObjects
+                }
+            });
+        }
+
+        /**
+         * Check if the given name is a restricted because '*' is being imported.
+         * @param {string} name - path of the import
+         * @param {node} node - representing the restricted path reference
+         * @returns {boolean} whether the path is restricted
+         * @private
+         */
+        function isRestrictedForEverythingImported(name, node) {
+            return Object.prototype.hasOwnProperty.call(restrictedPathMessages, name) &&
+                restrictedPathMessages[name].objectNames &&
+                isEverythingImported(node);
+        }
+
+        /**
+         * Check if the given objectName is restricted given a list of restrictedObjectNames.
+         * @param {[string]} objectNames - array of object names that are being imported
+         * @param {[string]} restrictedObjectNames - array of object names that are restricted for this import
+         * @returns {boolean} whether the objectName is restricted
+         * @private
+         */
+        function isRestrictedObject(objectNames, restrictedObjectNames) {
+            const restrictedObjectMatches = restrictedObjectNames.reduce((accumulator, restrictedObjectName) => {
```

Comment:
I think it would be better to do something like this to avoid quadratic time:

```js
function isRestrictedObject(objectNames, restrictedObjectNames) {
    const restrictedNamesSet = new Set(restrictedObjectNames);

    return objectNames.some(name => restrictedNamesSet.has(name));
}
```

Alternatively, you could use something like [`lodash.intersection`](https://lodash.com/docs/4.17.4#intersection).

Reply:
Great suggestion.  I did the inverse of this and changed `objectNames` (now `importNames`) to be build as a set instead of an array.  

I have changed this to be:

```
function isRestrictedObject(importNames, restrictedImportNames) {
    return restrictedImportNames.some(restrictedObjectName => (
        importNames.has(restrictedObjectName)
    ));
}
```

Let me know what you think.  These changes are in the fifth commit.